src/ceph-volume/ceph_volume/devices/lvm/listing.py

```diff
@@ -126,142 +115,87 @@ def list(self, args):
-            except MultipleLVsError:
-                lvs.filter(vg_name=pv.vg_name)
-                return self.full_report(lvs=lvs)
-
-        if lv:
-            try:
-                _id = lv.tags['ceph.osd_id']
-            except KeyError:
-                logger.warning('device is not part of ceph: %s', device)
-                return report
 
+        lvs = [lvs] if isinstance(lvs, api.Volume) else lvs
+        for lv in lvs:
+            if not is_ceph_device(lv):
+                continue
+
+            _id = lv.tags['ceph.osd_id']
             report.setdefault(_id, [])
+
             lv_report = lv.as_dict()
-            lv_report['devices'] = self.match_devices(lv.lv_uuid)
-            report[_id].append(lv_report)
+            pvs = api.get_pvs(filters='lv_uuid=' + lv.lv_uuid)
+            lv_report['devices'] = [pv.name for pv in pvs] if pvs else []
+
+            device_type = get_device_type(lv)
+            if device_type == 'lv':
+                report[_id].append(lv_report)
+            else:
+                create_report_for_nonlv_devs(lv, report)
+
+            if full_report:
+                for device_type in ['journal', 'block', 'wal', 'db']:
+                    create_report_for_nonlv_devs(lv, report)
 
-        else:
-            # this has to be a journal/wal/db device (not a logical volume) so try
-            # to find the PARTUUID that should be stored in the OSD logical
-            # volume
-            for device_type in ['journal', 'block', 'wal', 'db']:
-                device_tag_name = 'ceph.%s_device' % device_type
-                device_tag_uuid = 'ceph.%s_uuid' % device_type
-                associated_lv = lvs.get(lv_tags={device_tag_name: device})
-                if associated_lv:
-                    _id = associated_lv.tags['ceph.osd_id']
-                    uuid = associated_lv.tags[device_tag_uuid]
-
-                    report.setdefault(_id, [])
-                    report[_id].append(
-                        {
-                            'tags': {'PARTUUID': uuid},
-                            'type': device_type,
-                            'path': device,
-                        }
-                    )
         return report
 
-    def full_report(self, lvs=None):
-        """
-        Generate a report for all the logical volumes and associated devices
-        that have been previously prepared by Ceph
+    def full_report(self):
         """
-        if lvs is None:
-            lvs = api.Volumes()
-        report = {}
+        Create a report of all Ceph LVs.
 
-        for lv in lvs:
-            try:
-                _id = lv.tags['ceph.osd_id']
-            except KeyError:
-                # only consider ceph-based logical volumes, everything else
-                # will get ignored
-                continue
+        Must return '{}' to denote failure.
+        """
+        return self.create_report(api.Volumes(), full_report=True)
```

Comment:
Let's not use `Volumes` here. The point of this refactoring is to get rid of these containers.

Reply:
Got rid of it.